android-sdk/README.md

```diff
@@ -0,0 +1,224 @@
+    private static final Map<String, String> dependencyMap = new HashMap<>();
+    /**
+     * Register the looking for a service module to a class.
+     *
+     * @param type the type of module to use
+     * @param moduleClass the class that implements a service module type
+     * @param dependsOn the other serviceTypes this Service Depends on.
+     */
+    public static void registerServiceModule(String type, Class<? extends ServiceModule> moduleClass, String... dependsOn) {
+        serviceTypeMap.put(type, moduleClass);
+        for (String dependency : dependsOn) {
+            dependencyMap.put(type, dependency);
+        }
+    }
+
+}
+```
+** Discussion Point ** I'm not a fan of static methods during testing.  Perhaps there should be a method in the ServiceModule interface which returns a list of services that the service depends on.  MobileCore could then read that list by invoking that method as part of bootstrapping.  Furthermore, if we move to using Dagger to manage service modules it will elminate the static problem but then we will need to have constructor based injection.  Therefore in a future implementation ServiceRegistry will probably be implemented by or replaced with Dagger. This will be described more fully in a future proposal.
+
+#### Referencing a Service
+An instance of the core bassed during bootstrap can be used to reference other services.  MobileCore is responsible for ensuring that ServiceModule.bootstrap is not called until all declared dependencies are bootstrapped.
+
+SyncService.java
+```
+public class SyncService implements ServiceModule {
+
+  KeyCloakService keycloakService;
+  /*snip*/
+  public void bootstrap(MobileCore core, ServiceConfiguration config) {
+     this.keycloakService = core.getService("keycloak", KeyCloakService.class);
+     this.networkService = core.getService("network", NetworkService.class);
+     /*snip*/
+  }
+
+  public void performSync() {
+    if (keycloakService.isLoggedIn()) {
+      Reuqest request = networkService.newRequest();
+      request.addSecurityHeaders(keycloakService.getHeaders());//pseudocode method, don't expect this exactly.
+      request.setData(this.buildrequestBody());
+      Response response = networkService.execute(request);
+      response.onError(/*Error handling lambda*/);
+      response.onSuccess(/*Success handling lambda*/);
+    }
+  }
+
+}
+```
+
+#### Using Core Threads
+
+The Mobile Core SDK will provide Java Executors that wrap several common blocking tasks.  These executors will be available to any module or part of the application which may need to schedule diskIO work, netowrking work, or work to happen on the main thread.
+
+```
+AppExecutors.networkThread.execute(() -> {
+  var result = someRequest.call(); //Perform a HTTP operation off the main thread
+  AppExecutors.mainThread.execute(()->{updateDisplay(result);});//Update the Android UI on the main thread.
+ });
+```
+
+#### Using Core Networking
```

Comment:
Are we planning to implement out entire network library? Like our own Retrofit/Volley kind of thing?

Reply:
That is a debate for the future.

Ideally we will wrap our choice of implementation in a API so we can make behavior changes under the hood.

Reply:
OK that make more sense